File: products/compliance/src/reporters/html.ts
import type { ScanResult, Severity, Framework, Finding } from '../types.js';
import { calculateScore, scoreColor } from './badge.js';

const severityOrder: Severity[] = ['critical', 'high', 'medium', 'low', 'info'];

const severityColorMap: Record<Severity, string> = {
  critical: '#e05d44',
  high: '#d9534f',
  medium: '#f0ad4e',
  low: '#5bc0de',
  info: '#999',
};

function escapeHtml(s: string): string {
  return s
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;');
}

function severityBarChart(bySeverity: Record<Severity, number>): string {
  const maxCount = Math.max(...Object.values(bySeverity), 1);
  const barWidth = 200;
  const barHeight = 24;
  const gap = 6;
  const labelWidth = 70;
  const countWidth = 40;
  const totalHeight = severityOrder.length * (barHeight + gap);

  const bars = severityOrder
    .map((sev, i) => {
      const count = bySeverity[sev];
      const w = Math.max((count / maxCount) * barWidth, 0);
      const y = i * (barHeight + gap);
      const color = severityColorMap[sev];
      return `
      <g transform="translate(0,${y})">
        <text x="${labelWidth - 4}" y="${barHeight / 2 + 5}" text-anchor="end" font-size="13" fill="#444">${sev.toUpperCase()}</text>
        <rect x="${labelWidth}" y="0" width="${w}" height="${barHeight}" rx="3" fill="${color}" />
        <text x="${labelWidth + w + 6}" y="${barHeight / 2 + 5}" font-size="13" fill="#333">${count}</text>
      </g>`;
    })
    .join('');

  return `<svg width="${labelWidth + barWidth + countWidth}" height="${totalHeight}" xmlns="http://www.w3.org/2000/svg" role="img" aria-label="Severity distribution chart">${bars}</svg>`;
}

function frameworkSection(findings: Finding[], frameworks: Framework[]): string {
  const rows = frameworks
    .map((fw) => {
      const fwFindings = findings.filter((f) => f.framework.includes(fw));
      const count = fwFindings.length;
      return `<tr><td>${fw.toUpperCase()}</td><td>${count}</td></tr>`;
    })
    .join('');
  return `
    <table class="fw-table">
      <thead><tr><th>Framework</th><th>Findings</th></tr></thead>
      <tbody>${rows}</tbody>
    </table>`;
}

function findingsTable(findings: Finding[], severity: Severity): string {
  const filtered = findings.filter((f) => f.severity === severity);
  if (filtered.length === 0) return '';

  const rows = filtered
    .map(
      (f) => `
      <tr>
        <td>${escapeHtml(f.id)}</td>
        <td>${escapeHtml(f.title)}</td>
        <td>${f.file ? escapeHtml(f.file) + (f.line ? ':' + f.line : '') : 'project-level'}</td>
        <td>${f.framework.map((fw) => fw.toUpperCase()).join(', ')}</td>
        <td>${f.fixable ? 'Yes' : 'No'}</td>
      </tr>`,
    )
    .join('');

  const color = severityColorMap[severity];

  return `
    <details open>
      <summary style="color:${color};font-weight:bold;font-size:1.1em;cursor:pointer;">
        ${severity.toUpperCase()} (${filtered.length})
      </summary>
      <table class="findings-table">
        <thead>
          <tr><th>ID</th><th>Title</th><th>Location</th><th>Frameworks</th><th>Fixable</th></tr>
        </thead>
        <tbody>${rows}</tbody>
      </table>
    </details>`;
}

/**
 * Generate a standalone HTML compliance report with embedded CSS.
 *
 * Sections:
 *  - Executive summary (score, scan date, framework coverage)
 *  - Findings by severity (collapsible)
 *  - Findings by framework
 *  - Inline SVG severity distribution chart
 *  - Soul branding header/footer
 */
export function generateHtml(result: ScanResult, totalRules: number): string {
  const score = calculateScore(result.summary.total, totalRules);
  const color = scoreColor(score);
  const scanDate = result.metadata.timestamp;
  const frameworkList = result.metadata.frameworks.map((f) => f.toUpperCase()).join(', ');

  const severitySections = severityOrder
    .map((sev) => findingsTable(result.findings, sev))
    .join('');

  const chart = severityBarChart(result.summary.bySeverity);

  const fwSection = frameworkSection(result.findings, result.metadata.frameworks);

  return `<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8" />
<meta name="viewport" content="width=device-width, initial-scale=1.0" />
<title>Soul Compliance Report</title>
<style>
  * { box-sizing: border-box; margin: 0; padding: 0; }
  body { font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif; color: #333; background: #f9fafb; line-height: 1.6; }
  .header { background: #1a1a2e; color: #fff; padding: 24px 32px; display: flex; align-items: center; justify-content: space-between; }
  .header h1 { font-size: 1.6em; font-weight: 600; }
  .header .brand { font-size: 0.9em; opacity: 0.8; }
  .container { max-width: 960px; margin: 0 auto; padding: 24px; }
  .executive-summary { background: #fff; border-radius: 8px; padding: 24px; margin-bottom: 24px; box-shadow: 0 1px 3px rgba(0,0,0,0.08); }
  .executive-summary h2 { margin-bottom: 16px; font-size: 1.3em; }
  .score-badge { display: inline-block; font-size: 2em; font-weight: 700; padding: 8px 20px; border-radius: 8px; color: #fff; }
  .meta-grid { display: grid; grid-template-columns: 1fr 1fr; gap: 12px; margin-top: 16px; }
  .meta-grid dt { font-weight: 600; color: #555; }
  .meta-grid dd { margin: 0 0 8px 0; }
  .section { background: #fff; border-radius: 8px; padding: 24px; margin-bottom: 24px; box-shadow: 0 1px 3px rgba(0,0,0,0.08); }
  .section h2 { margin-bottom: 16px; font-size: 1.2em; }
  details { margin-bottom: 12px; }
  summary { padding: 8px 0; }
  table { width: 100%; border-collapse: collapse; margin-top: 8px; }
  th, td { text-align: left; padding: 8px 12px; border-bottom: 1px solid #eee; font-size: 0.9em; }
  th { background: #f5f5f5; font-weight: 600; }
  .findings-table { margin-bottom: 16px; }
  .fw-table { max-width: 400px; }
  .chart-wrapper { display: flex; justify-content: center; padding: 16px 0; }
  .footer { text-align: center; padding: 24px; color: #888; font-size: 0.85em; border-top: 1px solid #eee; margin-top: 32px; }
</style>
</head>
<body>
<div class="header">
  <h1>Soul Compliance Report</h1>
  <span class="brand">Powered by Soul</span>
</div>
<div class="container">
  <div class="executive-summary">
    <h2>Executive Summary</h2>
    <span class="score-badge" style="background:${color}">${score}%</span>
    <dl class="meta-grid">
      <dt>Scan Date</dt>
      <dd>${escapeHtml(scanDate)}</dd>
      <dt>Directory</dt>
      <dd>${escapeHtml(result.metadata.directory)}</dd>
      <dt>Frameworks</dt>
      <dd>${frameworkList}</dd>
      <dt>Total Findings</dt>
      <dd>${result.summary.total}</dd>
      <dt>Total Rules</dt>
      <dd>${totalRules}</dd>
      <dt>Duration</dt>
      <dd>${result.metadata.duration}ms</dd>
      <dt>Analyzers</dt>
      <dd>${result.metadata.analyzersRun.join(', ')}</dd>
      <dt>Auto-fixable</dt>
      <dd>${result.summary.fixable}</dd>
    </dl>
  </div>

  <div class="section">
    <h2>Severity Distribution</h2>
    <div class="chart-wrapper">${chart}</div>
  </div>

  <div class="section">
    <h2>Findings by Severity</h2>
    ${severitySections || '<p>No findings.</p>'}
  </div>

  <div class="section">
    <h2>Findings by Framework</h2>
    ${fwSection}
  </div>
</div>
<div class="footer">
  Generated by Soul Compliance &mdash; ${escapeHtml(scanDate)}
</div>
</body>
</html>`;
}

File: products/compliance/reporters/html.go
package reporters

import (
	"fmt"
	"html"
	"strings"

	"github.com/rishav1305/soul/products/compliance/scan"
)

// GenerateHTML returns a standalone HTML page containing the scan results
// with inline CSS styling, a summary table, findings table, and score display.
func GenerateHTML(result *scan.ScanResult) string {
	score := CalculateScore(result)
	scoreColor := "#e05d44"
	switch {
	case score >= 90:
		scoreColor = "#4c1"
	case score >= 70:
		scoreColor = "#dfb317"
	case score >= 50:
		scoreColor = "#fe7d37"
	}

	var b strings.Builder

	b.WriteString(`<!DOCTYPE html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Compliance Scan Report</title>
<style>
  * { margin: 0; padding: 0; box-sizing: border-box; }
  body { font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif; background: #f5f5f5; color: #333; padding: 2rem; }
  .container { max-width: 1200px; margin: 0 auto; }
  h1 { color: #1a1a2e; margin-bottom: 1rem; }
  h2 { color: #16213e; margin: 1.5rem 0 0.5rem; }
  .score-badge { display: inline-block; padding: 0.5rem 1.5rem; border-radius: 8px; color: white; font-size: 1.5rem; font-weight: bold; margin: 1rem 0; }
  .meta { background: white; border-radius: 8px; padding: 1rem 1.5rem; margin-bottom: 1.5rem; box-shadow: 0 1px 3px rgba(0,0,0,0.1); }
  .meta p { margin: 0.25rem 0; color: #555; }
  .meta strong { color: #333; }
  table { width: 100%; border-collapse: collapse; background: white; border-radius: 8px; overflow: hidden; box-shadow: 0 1px 3px rgba(0,0,0,0.1); margin-bottom: 1.5rem; }
  th { background: #1a1a2e; color: white; padding: 0.75rem 1rem; text-align: left; font-weight: 600; }
  td { padding: 0.75rem 1rem; border-bottom: 1px solid #eee; }
  tr:last-child td { border-bottom: none; }
  tr:hover td { background: #f8f9fa; }
  .severity-critical { color: #e05d44; font-weight: bold; }
  .severity-high { color: #dfb317; font-weight: bold; }
  .severity-medium { color: #0e7490; font-weight: bold; }
  .severity-low { color: #666; }
  .severity-info { color: #999; }
  .fixable { color: #4c1; font-size: 0.85rem; }
  .evidence { font-family: monospace; font-size: 0.85rem; color: #666; max-width: 300px; overflow: hidden; text-overflow: ellipsis; white-space: nowrap; }
  .summary-grid { display: grid; grid-template-columns: repeat(auto-fit, minmax(150px, 1fr)); gap: 1rem; margin-bottom: 1.5rem; }
  .summary-card { background: white; border-radius: 8px; padding: 1rem; text-align: center; box-shadow: 0 1px 3px rgba(0,0,0,0.1); }
  .summary-card .count { font-size: 2rem; font-weight: bold; }
  .summary-card .label { font-size: 0.85rem; color: #666; text-transform: uppercase; }
  footer { text-align: center; color: #999; margin-top: 2rem; font-size: 0.85rem; }
</style>
</head>
<body>
<div class="container">
`)

	// Header and score
	b.WriteString(`<h1>Compliance Scan Report</h1>`)
	b.WriteString(fmt.Sprintf(`<div class="score-badge" style="background:%s">Score: %d%%</div>`,
		scoreColor, score))

	// Metadata
	b.WriteString(`<div class="meta">`)
	b.WriteString(fmt.Sprintf(`<p><strong>Directory:</strong> %s</p>`, html.EscapeString(result.Metadata.Directory)))
	b.WriteString(fmt.Sprintf(`<p><strong>Duration:</strong> %.2fs</p>`, result.Metadata.Duration))
	b.WriteString(fmt.Sprintf(`<p><strong>Analyzers:</strong> %s</p>`, html.EscapeString(strings.Join(result.Metadata.AnalyzersRun, ", "))))
	b.WriteString(fmt.Sprintf(`<p><strong>Timestamp:</strong> %s</p>`, html.EscapeString(result.Metadata.Timestamp)))
	b.WriteString(`</div>`)

	// Summary cards
	b.WriteString(`<h2>Summary</h2>`)
	b.WriteString(`<div class="summary-grid">`)
	b.WriteString(fmt.Sprintf(`<div class="summary-card"><div class="count">%d</div><div class="label">Total</div></div>`, result.Summary.Total))
	b.WriteString(fmt.Sprintf(`<div class="summary-card"><div class="count">%d</div><div class="label">Fixable</div></div>`, result.Summary.Fixable))

	for _, sev := range severityOrder {
		count := result.Summary.BySeverity[sev]
		if count == 0 {
			continue
		}
		b.WriteString(fmt.Sprintf(`<div class="summary-card"><div class="count severity-%s">%d</div><div class="label">%s</div></div>`,
			sev, count, strings.ToUpper(sev)))
	}
	b.WriteString(`</div>`)

	// Findings table
	if result.Summary.Total > 0 {
		b.WriteString(`<h2>Findings</h2>`)
		b.WriteString(`<table>`)
		b.WriteString(`<thead><tr><th>ID</th><th>Severity</th><th>Title</th><th>File</th><th>Line</th><th>Evidence</th><th>Fixable</th></tr></thead>`)
		b.WriteString(`<tbody>`)

		for _, f := range result.Findings {
			fixLabel := ""
			if f.Fixable {
				fixLabel = `<span class="fixable">Yes</span>`
			}
			location := html.EscapeString(f.File)
			lineStr := ""
			if f.Line > 0 {
				lineStr = fmt.Sprintf("%d", f.Line)
			}

			b.WriteString(fmt.Sprintf(`<tr><td>%s</td><td class="severity-%s">%s</td><td>%s</td><td>%s</td><td>%s</td><td class="evidence">%s</td><td>%s</td></tr>`,
				html.EscapeString(f.ID),
				strings.ToLower(f.Severity),
				html.EscapeString(strings.ToUpper(f.Severity)),
				html.EscapeString(f.Title),
				location,
				lineStr,
				html.EscapeString(f.Evidence),
				fixLabel,
			))
		}

		b.WriteString(`</tbody></table>`)
	}

	// Footer
	b.WriteString(`<footer>Generated by compliance-go</footer>`)
	b.WriteString(`</div></body></html>`)

	return b.String()
}
